File: src/commands/shell.rs
use crate::calm_format;
use crate::config;
use crate::theme_engine::Theme;
use anyhow::Result;

/// Entry point for bare `calm` — launches the interactive shell.
///
/// Renders the branded banner (config.calm-driven), then hands off to
/// the real line editor (see `line_editor.rs` / `repl.rs`) for the
/// interactive loop: autosuggestions, syntax highlighting, aliases,
/// functions/plugins, the `cd` builtin, and history.
pub fn launch() -> Result<()> {
    config::ensure_scaffold()?;
    let theme = Theme::load_active()?;

    // config.calm drives whether the banner/greeting shows and lets the
    // prompt icon be overridden per-user without touching the theme file.
    // A parse error anywhere in the include chain (config.calm or any
    // topic file it includes) used to fall back to a blank, empty config
    // with zero indication anything was wrong — every alias, function,
    // and setting the user wrote would silently vanish, banner and all,
    // with the shell looking perfectly normal. Loud beats silent here:
    // the shell still starts (a broken config shouldn't mean no shell at
    // all), but the user finds out immediately, not by wondering days
    // later why `..` stopped working.
    let cfg = match calm_format::parse_file(&config::config_file()?) {
        Ok(cfg) => cfg,
        Err(e) => {
            eprintln!("calm: could not load config, starting with defaults instead: {e}");
            calm_format::CalmDocument::default()
        }
    };
    let greeting = cfg.get_bool("shell", "greeting").unwrap_or(true);
    let icon_override = cfg.get_str("prompt", "icon").map(str::to_string);
    let icon: &str = icon_override.as_deref().unwrap_or_else(|| theme.icon());

    if greeting {
        println!("{}  {}", icon, theme.paint("calm_purple", "Calm-shell"));
        println!("{}", theme.paint("cloud_gray", &format!("theme: {}", theme.display_name())));
        println!();
    }

    crate::repl::run(&theme, &cfg, icon)
}

File: src/commands/shell.c
/* Renders the branded banner (config.calm-driven), then hands off to
 * the real line editor / REPL for the interactive loop. */
#include "shell.h"

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../calm_format.h"
#include "../config.h"
#include "../repl.h"
#include "../theme.h"
#include "../util.h"

int shell_cmd_launch(void) {
    if (!ensure_scaffold()) {
        fprintf(stderr, "error: could not set up config directory\n");
        return 1;
    }

    Theme theme;
    if (!theme_load_active(&theme)) {
        fprintf(stderr, "error: could not load any theme (not even the bundled calm-lavender)\n");
        return 1;
    }

    /* config.calm drives whether the banner/greeting shows and lets the
     * prompt icon be overridden per-user without touching the theme
     * file. A parse error anywhere in the include chain used to fall
     * back to a blank, empty config with zero indication anything was
     * wrong -- every alias, function, and setting the user wrote would
     * silently vanish. Loud beats silent here: the shell still starts,
     * but the user finds out immediately. */
    char *config_path = config_file();
    CalmDocument cfg;
    char *err = NULL;
    bool cfg_ok = config_path && calm_parse_file(config_path, &cfg, &err);
    free(config_path);
    if (!cfg_ok) {
        fprintf(stderr, "calm: could not load config, starting with defaults instead: %s\n",
                err ? err : "unknown error");
        free(err);
        calm_document_init(&cfg);
    }

    bool greeting = calm_document_get_bool(&cfg, "shell", "greeting", true);
    const char *icon_override = calm_document_get_str(&cfg, "prompt", "icon");
    const char *icon = icon_override ? icon_override : theme.icon;

    if (greeting) {
        char *purple = theme_paint(&theme, "calm_purple", "Calm-shell");
        printf("%s  %s\n", icon, purple);
        free(purple);

        char *label = xsprintf("theme: %s", theme.display_name);
        char *painted_label = theme_paint(&theme, "cloud_gray", label);
        free(label);
        printf("%s\n\n", painted_label);
        free(painted_label);
    }

    int result = repl_run(&theme, &cfg, icon);

    calm_document_free(&cfg);
    theme_free(&theme);
    return result;
}
